File: 2019 카카오 개발자 겨울 인턴십/불량 사용자.cpp
#include <iostream>
#include <string>
#include <vector>
#include <set>

using namespace std;

const int MAX_USER_ID_LENGTH = 8;

set<int> s;
bool check[MAX_USER_ID_LENGTH];

void dfs(const vector<string>& user_id, const vector<string>& banned_id, int idx, int candi) {
    if(idx == banned_id.size()) {
        s.insert(candi);
        return;
    }
    for(int i = 0; i < user_id.size(); ++i) {
        if(user_id[i].size() != banned_id[idx].size())
            continue;
        if(check[i] == true)
            continue;
        int j;
        for(j = 0; j < user_id[i].size(); ++j) {
            if(banned_id[idx][j] == '*')
                continue;
            if(user_id[i][j] != banned_id[idx][j])
                break;
        }
        if(j == user_id[i].size()) {
            check[i] = true;
            dfs(user_id, banned_id, idx+1, candi | (1 << i));
            check[i] = false;
        }
    }
}

int solution(vector<string> user_id, vector<string> banned_id) {
    dfs(user_id, banned_id, 0, 0);
    return s.size();
}

File: 2019 카카오 개발자 겨울 인턴십/불량 사용자.java
import java.util.*;

class Solution {
    static final int MAX_USER_ID_LENGTH = 8;
    
    String[] User_id, Banned_id;
    boolean[] check = new boolean[MAX_USER_ID_LENGTH];
    Set<Integer> mySet = new HashSet<>();
    
    void findUserIdList(int index, int candidate) {
     	if(index == Banned_id.length) {
            mySet.add(candidate);
            return;
        }   
        for(int i = 0; i < User_id.length; ++i) {
            if(User_id[i].length() != Banned_id[index].length())
                continue;
            if(check[i] == true)
                continue;
            int j;
            for(j = 0; j < User_id[i].length(); ++j) {
                if(Banned_id[index].charAt(j) == '*')
                    continue;
                if(User_id[i].charAt(j) != Banned_id[index].charAt(j))
                    break;
            }
            if(j == User_id[i].length()) {
                check[i] = true;
                findUserIdList(index+1, candidate | (1 << i));
                check[i] = false;
            }
        }
    }
    
    public int solution(String[] user_id, String[] banned_id) {
        User_id = user_id;
        Banned_id = banned_id;
        
        findUserIdList(0, 0);
        
        return mySet.size();
    }
}
